[mlir][transform] Improve error message of tracking listener.

ingomueller-net opened the pull request:
This PR extends the error message of the tracking listener when replacement ops cannot be found. That may happen if the applied patterns replace an op by an op of a different kind or by block arguments. However, this only matters if there are alive handles to the replaced op. The new error message mentions that explicitly and reports the alive handles.

ingomueller-net committed:
[mlir][transform] Improve error message of tracking listener.

This PR extends the error message of the tracking listener when
replacement ops cannot be found. That may happen if the applied patterns
replace an op by an op of a different kind or by block arguments.
However, this only matters if there are alive handles to the replaced
op. The new error message mentions that explicitly and reports the alive
handles.

ftynse commented on mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp:
```diff
@@ -1396,14 +1396,13 @@ void transform::TrackingListener::notifyOperationReplaced(
-    }
-    return false;
-  };
+  SmallVector<Operation *> aliveUsers;
```
In theory, we can have a lot of users and reporting all of them isn't necessarily a readability improvement as even a dozen of notes will fill up the terminal window. How about reporting the _first found_ user instead?

matthias-springer replied:
Yes I think the important piece of information here is that the is at least one user, i.e., the handle is not dead. It is then pretty easy to find that user when looking at the transform script.

ftynse commented on mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp:
```diff
@@ -1444,16 +1443,20 @@ bool transform::ErrorCheckingTrackingListener::failed() const {
   if (status.succeeded()) {
     status = emitSilenceableFailure(
-        getTransformOp(), "tracking listener failed to find replacement op");
+        getTransformOp(), "op was replaced but replacement was of different "
+                          "kind, invalidating alive handles");
```
Replacement may not be found for other reasons though, like replacement values belonging to different ops (so there is no _single_ replacement though all of the new producers have the right type). I suspect the message should be produced by the function that looks for replacement and forwarded here.

ftynse replied:
You can always return additional things via references, use `std::tuple` or implement `DiagnosedSilenceableFailureOr`.

ftynse commented on mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp:
```diff
@@ -1444,16 +1443,20 @@ bool transform::ErrorCheckingTrackingListener::failed() const {
         << "[" << errorCounter << "] replacement value " << index;
+  for (auto &&[index, user] : llvm::enumerate(aliveUsers))
+    status.attachNote(user->getLoc())
+        << "[" << errorCounter << "] alive handle " << index;
```
This feels misleading. The user is not the handle, its _operand_ is. We should capture uses (`OpOperand *`) rather than users and attach the note to the location of the value that is being used. Bonus points for giving extra information like the value being n-th result of the op (values have the location of the defining op so we cannot otherwise differentiate between them).

matthias-springer commented on mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp:
```diff
@@ -1444,16 +1443,20 @@ bool transform::ErrorCheckingTrackingListener::failed() const {
+    Operation *op, ValueRange values, ArrayRef<Operation *> aliveUsers) {
   if (status.succeeded()) {
     status = emitSilenceableFailure(
-        getTransformOp(), "tracking listener failed to find replacement op");
```
maybe "tracking listener failed to find replacement op for op that is tracked by non-dead handle".

But even that could be confusion. Maybe we should just add a comment in the C++ code that mentions all the conditions (type changed, handle not dead, etc.).

ftynse replied:
I was thinking literally returning `DiagnosedSilecenableFailure` instead of `LogicalResult`, then the replacement lookup can specify the message.

ingomueller-net committed:
Address @ftynse's and @matthias-springer's comments.

In particular:

* Change `TrackingListener::findReplacementOp` such that it returns a
  `DiagnosedSilenceableFailure` rather than just a `FailureOr`.
* Have `TrackingListener::notifyPayloadReplacementNotFound` accept a
  `DiagnosedSilenceableFailure` with previous diagnostics.
* Create initial diagnostics in `findReplacementOp` and give information
  about why finding the replacement failed.
* Forward that result into `notifyPayloadReplacementNotFound` and
  adapt how the `ErroriCheckingTrackingListener` deals with the
  pre-existing diagnostics.
* Only report the first alive user.
* Adapt error messages to suggestions.

ingomueller-net commented on mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp:
```diff
@@ -1444,16 +1443,20 @@ bool transform::ErrorCheckingTrackingListener::failed() const {
+                          "kind, invalidating alive handles");
   }
 
   status.attachNote(op->getLoc()) << "[" << errorCounter << "] replaced op";
   for (auto &&[index, value] : llvm::enumerate(values))
     status.attachNote(value.getLoc())
         << "[" << errorCounter << "] replacement value " << index;
```
This information could (almost) be added to the diagnostics `findReplacementOp`; however, `errorCounter` is not available there. Are we sure we need this? (There are no tests where `errorCounter > 0`.) Note that some of the diagnostics that this PR currently adds also don't report the current error count...

ingomueller-net reviewed (commented):
Thanks a lot for the reviews! Issues addressed (or attempted to do so). Please have another look!

ingomueller-net committed:
Fix test for tensor transform ops.

That op created an empty (and invalid) transform op with
`TransformOpInterface()` for testing the tracking listeners. The
previous listeners never accessed the op created this way, so this
wasn't a problem. However, the new version does, so we need to create a
real op, which is what this commit does.

ftynse reviewed (approved):
Nice!

ingomueller-net commented:
Thanks for notifying! Saw it just now as well. I am AFK and can only fix tomorrow and revert in 30m-2h. I'll do that if nobody else has reverted by then.

vitalybuka commented:
Thanks, I will revert.

vitalybuka commented:
Reverted 78c8ab5844e618162c4cf3982d05102d4da10d23

matthias-springer commented on mlir/test/lib/Dialect/Tensor/TestTensorTransforms.cpp:
```diff
@@ -352,8 +352,17 @@ static LogicalResult testTrackingListenerReplacements(Operation *rootOp) {
-  DummyTrackingListener listener(transformState,
-                                 transform::TransformOpInterface());
+  MLIRContext *context = rootOp->getContext();
+  OpBuilder builder(context);
```
This OpBuilder has no insertion point. That's likely what caused the leak.
